sql-plugin/src/main/scala/com/nvidia/spark/rapids/limit.scala

```diff
@@ -58,61 +58,75 @@ trait GpuBaseLimitExec extends LimitExec with GpuExec with ShimUnaryExecNode {
     throw new IllegalStateException(s"Row-based execution should not occur for $this")
 
   override def doExecuteColumnar(): RDD[ColumnarBatch] = {
-    val numOutputRows = gpuLongMetric(NUM_OUTPUT_ROWS)
-    val numOutputBatches = gpuLongMetric(NUM_OUTPUT_BATCHES)
-    val opTime = gpuLongMetric(OP_TIME)
+    sliceRDD(child.executeColumnar(), limit, 0)
+  }
 
-    val crdd = child.executeColumnar()
-    crdd.mapPartitions { cbIter =>
+  def sliceRDD(rdd: RDD[ColumnarBatch], limit: Int, offset: Int): RDD[ColumnarBatch] = {
+    val opTime = gpuLongMetric(OP_TIME)
+    rdd.mapPartitions { iter =>
       new Iterator[ColumnarBatch] {
-        private var remainingLimit = limit
+        private var remainingLimit = limit - offset
+        private var remainingOffset = offset
 
-        override def hasNext: Boolean = remainingLimit > 0 && cbIter.hasNext
+        override def hasNext: Boolean = iter.hasNext
 
         override def next(): ColumnarBatch = {
-          val batch = cbIter.next()
-          withResource(new NvtxWithMetrics("limit", NvtxColor.ORANGE, opTime)) { _ =>
-            val result = if (batch.numRows() > remainingLimit) {
-              sliceBatch(batch)
+          var batch = iter.next()
+          val numCols = batch.numCols()
+
+          // In each partition, we need to skip `offset` rows
+          while (batch != null && remainingOffset >= batch.numRows()) {
+            remainingOffset -= batch.numRows()
+            batch.safeClose()
+            batch = if (hasNext) { iter.next() } else { null }
+          }
+
+          // If the last batch is null, then we have offset >= numRows in this partition.
+          // In such case, we should return an empty batch
+          if (batch == null) {
+            return new ColumnarBatch(new ArrayBuffer[GpuColumnVector](numCols).toArray, 0)
+          }
+
+          // Here remainingOffset < batch.numRow(), we need to get batch[remainingOffset:]
+          withResource(new NvtxWithMetrics("limit and offset", NvtxColor.ORANGE, opTime)) { _ =>
+            var result: ColumnarBatch = null
+            // If limit != -1, then remainingLimit >= 0 is valid, otherwise remainingLimit < 0,
+            // we should take batch[remainingOffset:numRows].
+            if (remainingLimit >= 0) {
```

Comment:
Per the updated `hasNext`, `remainingLimit` can not be equal to 0. 
Even if `remainingLimit == 0`, we should return an empty batch.